File: frontend/src/api/predictions.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiGet, apiPost } from "./client";

export type RiskLevel = "stable" | "at_risk" | "unstable";

export interface CurvePoint {
  pressure: number;
  instability_index: number;
}

export interface PredictionInput {
  well_id: number;
  reservoir_pressure_psi: number;
  reservoir_temperature_f: number;
  api_gravity: number;
  gas_specific_gravity: number;
  solution_gor_scf_stb: number;
  resin_asphaltene_ratio: number;
}

export interface Prediction extends PredictionInput {
  id: number;
  bubble_point_pressure_psi: number;
  onset_pressure_psi: number;
  risk_level: RiskLevel;
  curve: CurvePoint[];
  created_at: string;
}

export interface PredictionSummary {
  id: number;
  well_id: number;
  risk_level: RiskLevel;
  onset_pressure_psi: number;
  bubble_point_pressure_psi: number;
  created_at: string;
}

const PREDICTIONS_QUERY_KEY = ["predictions"] as const;

export function usePredictionsQuery(wellId: number | undefined) {
  return useQuery({
    queryKey: [...PREDICTIONS_QUERY_KEY, wellId],
    queryFn: () =>
      apiGet<{ items: PredictionSummary[]; total: number }>(`/predictions?well_id=${wellId}`),
    enabled: wellId !== undefined,
  });
}

export function usePredictionQuery(predictionId: number | undefined) {
  return useQuery({
    queryKey: [...PREDICTIONS_QUERY_KEY, "detail", predictionId],
    queryFn: () => apiGet<Prediction>(`/predictions/${predictionId}`),
    enabled: predictionId !== undefined,
  });
}

export function useCreatePredictionMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: PredictionInput) => apiPost<Prediction>("/predictions", input),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: PREDICTIONS_QUERY_KEY }),
  });
}

File: backend/app/schemas/predictions.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.prediction import RiskLevel


class PredictionInput(BaseModel):
    well_id: int
    reservoir_pressure_psi: float = Field(gt=0)
    reservoir_temperature_f: float = Field(gt=32, lt=400)
    api_gravity: float = Field(gt=0, lt=100)
    gas_specific_gravity: float = Field(gt=0, lt=3)
    solution_gor_scf_stb: float = Field(gt=0)
    resin_asphaltene_ratio: float = Field(gt=0)


class CurvePoint(BaseModel):
    pressure: float
    instability_index: float


class PredictionOut(BaseModel):
    id: int
    well_id: int
    reservoir_pressure_psi: float
    reservoir_temperature_f: float
    api_gravity: float
    gas_specific_gravity: float
    solution_gor_scf_stb: float
    resin_asphaltene_ratio: float
    bubble_point_pressure_psi: float
    onset_pressure_psi: float
    risk_level: RiskLevel
    curve: list[CurvePoint]
    created_at: datetime

    model_config = {"from_attributes": True}


class PredictionSummaryOut(BaseModel):
    id: int
    well_id: int
    risk_level: RiskLevel
    onset_pressure_psi: float
    bubble_point_pressure_psi: float
    created_at: datetime

    model_config = {"from_attributes": True}


class PredictionListOut(BaseModel):
    items: list[PredictionSummaryOut]
    total: int
